### DISCUSSION TITLE
Improve dynamic Hooks performances

### DISCUSSION THREAD

**Discussion by PululuK** [PrestaShop/PrestaShop#33404] [Ideas]
## Is your feature request related to a problem?

Dynamics hooks can be sources of performance issues and major resource consumers.
Consider the following case : 

- I have ten modules hooked to: `actionObjectProductUpdateAfter`
- I have a massive product import module (@StoreCommander for example)
- Suppose that in this import module I update the names of the 3000 products
- The ten modules hooked to `actionObjectProductUpdateBefore` will be called 3000 times.
- So far, all is well ... this is the expected behavior !

### Disadvantages

My ten modules are called because the product has been updated, but these ten modules are not necessarily wanted to be called when the product name is changed, but rather when the reference or the quantity is changed  :  

- Ten modules are called 3000 times unnecessarily
- Modules can unnecessarily perform heavy actions

## Describe the solution you'd like

When the hooks are triggered we also need to send the list of changed fields
here for example : https://github.com/PrestaShop/PrestaShop/blob/develop/classes/ObjectModel.php#L816

```php

$hook_params =  [
   'object' => $this, 
   'fields_updateds' => $fields_updateds
];

Hook::exec('actionObjectUpdateAfter', $hook_params);
Hook::exec('actionObject' . $this->getFullyQualifiedName() . 'UpdateAfter', $hook_params);
```

On the modules, we can decide to carry out actions according to the modified fields.
Example of a module that only wants to be called when the default product category has changed 

```php

public function hookActionObjectProductUpdateAfter(array $params)
{
    if (in_array('id_category_default', $params['fields_updateds'])) {
       //Trigger extremely heavy actions
    }
}

```

I am available to discuss it if it is not very clear !
Thanks

[a comment by an automated account is not shown]

**Comment by khouloudbelguith** [comment]
Ping @PrestaShop/prestashop-product-team @PrestaShop/prestashop-core-developers

**Comment by PululuK** [comment]
ping @matks  @jolelievre  @eternoendless @mickaelandrieu

**Comment by kpodemski** [comment]
I'd like to follow this idea with yet another one, let us disable execution of dynamic hooks with something like:

`Configuration::set('PS_DISABLE_ACTION_HOOKS', 1);`

For the original idea by @PululuK - i like it and +1 for it, of course it requires more from developers but it's worth to have such a possibility directly in core

**Comment by matks** [comment]
> * Ten modules are called 3000 times unnecessarily

This will stay true, the 3000 times will be performed. However following your suggestion you aim to be able to use an early `return;` statement to make these 3000 calls less expensive.

> I'd like to follow this idea with yet another one, let us disable execution of dynamic hooks with something like:
> 
> `Configuration::set('PS_DISABLE_ACTION_HOOKS', 1);`

But this will shutdown _all_ action hooks. @PululuK usecase is rather "I know I am going to perform action A (action A is, here, a massive product import) and I know modules B, C and D do not care about this action so I'd like to silence them".

Following this idea, I believe @PululuK would rather wish to be able to "put on mute" an accurate list of hooks (so he can re-enable them at the end of the import process) ?

Note: is that possible, from a module A, to un-hook a module B ? that might be part of the solution

**Comment by kpodemski** [comment]
@matks 
> 

my idea is not to change behavior of solution suggested by @PululuK - it's just another idea for similar issue, sometimes we just want to avoid situation where all dynamic hooks like `actionSomething` are fired. This:

> This will stay true, the 3000 times will be performed. However following your suggestion you aim to be able to use an early return; statement to make these 3000 calls less expensive.

is what i think is ok as part one, but part two would be to have possibility to turn off them completely for a moment

> Note: is that possible, from a module A, to un-hook a module B ? that might be part of the solution

yes, but that's not a case, from what i understand @PululuK wants to perform actions inside the hooks of other modules, he believe that if we as a developers have proper tools, we can make it less expensive, for example if we have module for meta tags regeneration, we check if fields like product name, product description etc. are altered, if so, we use our code, if not, we just `return;`

my idea is different, i don't want to allow other modules (or core) use `action` hooks, that's why i'd love to see a configuration check in `Hook::exec`, i'll try to make PoC of this ;)

**Comment by eternoendless** [comment]
Regarding the original request: "I want to be able to tell which fields were updated when the hook is called", it'd certainly be a useful feature, but it's impossible to implement right now.

As far as I know, ObjectModel does not have field-level persistence status tracking (for lack of a better way to describe it). This means that when you update a model, we have no way to tell which or its properties has been updated, so they are all included in the update statement.

If model field writes had been were performed through setters instead of making the fields public for writing,  we would be able to track their status, perform partial updates, and inform which properties have been updated.

Unfortunately ObjectModel does not support that, so we're stuck.

Regarding the possibility to prevent ObjectModel from firing hooks, I think it's completely doable – even if a little dangerous. Me I would do that either at model level (eg. `SomeModel::disableHooks()`) or at Hook level (eg. `Hook::disableHooks()`).

**Comment by kpodemski** [comment]
My idea was to do something like:

`Configuration::set('DISABLE_DYNAMIC_HOOKS', 1)`

then perform some actions, then fire loadConfiguration, thanks to this dynamic hooks are only disabled for a period of time we need.

Of course this is something for advanced users and specific cases only, but I had to do it in many client projects, while importing new products, perform other heavy changes on catalog etc.

I'm not pushing for having it in the core tho, idea for FriendsOfPresta maybe 🤷🏼‍♂️
 
For the original idea, Puluk proposed a good idea with PoC, you say that you're stuck because of something. All you need to do is just document properly what an author already proposed. This is not a best solution, but it's something what we can work with, some of us we'll work with 1.7 for next 5 or more years.

**Comment by matks** [comment]
> My idea was to do something like:
> 
> `Configuration::set('DISABLE_DYNAMIC_HOOKS', 1)`

If you submit a PR to provide this feature I dont see any reason to reject it 😊 . if it's useful for you, it's useful for other people and people not interested into this setting do not have to use it.

> For the original idea, Puluk proposed a good idea with PoC, you say that you're stuck because of something. All you need to do is just document properly what an author already proposed. This is not a best solution, but it's something what we can work with, some of us we'll work with 1.7 for next 5 or more years.

If I got it right, what @eternoendless pointed out is that there is no way currently to compute the `fields_updated` part without doing it manually. If we were using an ORM like [Doctrine](https://www.doctrine-project.org/projects/orm.html) we would have been able to implement a system that would fill automatically the `field_updated`.

Without this, the only way to implement `fields_updated` is to modify each hook call manually and write the code that defines it. A lot longer to do 😅

**Comment by letwang** [comment]
`fields_updated` is a great optimization idea. In fact, we can think more broadly. Currently, Hooks are executed without any conditions. We need to preset some `conditions` for each Hook for users to configure and then decide whether to trigger a certain Some Hooks can greatly improve the performance of Hooks, and all PrestaShop users will cheer for this.

Note that not only `fields_updated`, the best thing is to provide `conditions` for users to configure!

In addition, when the Hook has not changed, you need to use inter-process caching to improve performance, instead of dynamically and repeatedly executing SQL every time. Solutions such as Redis, Yac, etc.

**Comment by Hlavtox** [comment]
@kpodemski What defines a dynamic hook? I can trigger any hook thousand of times with the correct scenario.

**Comment by kpodemski** [comment]
@Hlavtox 

for example: `actionObject<ObjectName>AddAfter`

but this topic is also related to a few other hooks like `actionProductUpdate`, `updateQuantity`, etc.

**Comment by PululuK** [comment]
Hi @matks @kpodemski 

Can you please check here : https://github.com/PrestaShop/PrestaShop/pull/37216
